Fix Usb Gadget when attached after power-up

IDisposable opened the pull request:
Fixes #128

### Summary
- Revamped the polling and writing of the USB gadget to correct for dropped connections
- Corrected the initialization order to ensure we run even if the USB is disconnected (and we're getting power from the y-splitter/POE)
- Ensured that we restart the LED listening polling
- If we exceed too many consecutive write errors, close and reopen the gadget

IDisposable commented on internal/usbgadget/log.go:
```diff
@@ -2,26 +2,55 @@ package usbgadget
+	// log if it's the first time, and then every N times thereafter
+	if !ok || counter%every == 0 {
+		logger.Error().Str("counterName", counterName).Int("counter", counter).Err(err).Msgf(msg, args...)
+		return ok // return whether we've just exceeded the every interval
```
No, we don't want to declare threshold exceeded when we _first_ start the counter.

IDisposable commented on usb_mass_storage.go:
```diff
@@ -41,13 +41,15 @@ func getMassStorageImage() (string, error) {
+
 	if err != nil {
-		return fmt.Errorf("failed to get mass storage path: %w", err)
+		return fmt.Errorf("failed to get mass storage path error: %w", err)
```
We don't have a path here, so we can't log it.

IDisposable commented on internal/usbgadget/hid_keyboard.go:
```diff
@@ -162,183 +186,169 @@ func (u *UsbGadget) SetOnKeepAliveReset(f func()) {
 }
 
-func (u *UsbGadget) cancelAutoRelease(key byte) {
+// note: lock must be freeed by caller
```
No, we don't want to release the lock inside this method. That's why we note that the caller must do so.

IDisposable commented on internal/usbgadget/hid_keyboard.go:
```diff
@@ -162,183 +186,169 @@ func (u *UsbGadget) SetOnKeepAliveReset(f func()) {
+			timer.Stop()
+		}
 
-	if timer := u.kbdAutoReleaseTimers[key]; timer != nil {
-		timer.Stop()
-		u.kbdAutoReleaseTimers[key] = nil
 		delete(u.kbdAutoReleaseTimers, key)
-
-		// Reset keep-alive timing when key is released
-		if u.onKeepAliveReset != nil {
-			(*u.onKeepAliveReset)()
-		}
 	}
+
+	return ok
 }
```
No, we do not want the lock released in the pop method. We want it released by the callers... hence the comment.

IDisposable commented on internal/usbgadget/hid_keyboard.go:
```diff
@@ -162,183 +186,169 @@ func (u *UsbGadget) SetOnKeepAliveReset(f func()) {
+func (u *UsbGadget) DelayAutoReleaseWithDuration(resetDuration time.Duration) {
+	logger := u.log.With().Dur("reset_duration", resetDuration).Logger()
+
 	u.kbdAutoReleaseLock.Lock()
-	defer unlockWithLog(&u.kbdAutoReleaseLock, u.log, "autoRelease scheduled")
+	defer logging.UnlockWithTraceLog(&u.kbdAutoReleaseLock, &logger, "auto-release delayed")
 
-	if u.kbdAutoReleaseTimers[key] != nil {
-		u.kbdAutoReleaseTimers[key].Stop()
+	for _, timer := range u.kbdAutoReleaseTimers {
+		if timer != nil {
+			timer.Reset(resetDuration)
+		}
 	}
```
No, we acquire the main lock before any setting/clearing/resetting of timers.

IDisposable commented on internal/usbgadget/hid_mouse_relative.go:
```diff
@@ -58,20 +58,17 @@ var relativeMouseCombinedReportDesc = []byte{
 	if err != nil {
-		u.logWithSuppression("relMouseWriteHidFile", 100, u.log, err, "failed to write to hidg2")
 		u.relMouseHidFile.Close()
 		u.relMouseHidFile = nil
```
This is by design, as  an attempt to recover if the file cannot be written. Since these are just HID Gadget writes, the loss of a write will be recovered shortly.

IDisposable commented on internal/usbgadget/usbgadget.go:
```diff
@@ -172,6 +170,12 @@ func (u *UsbGadget) Close() error {
+
 	// Close HID files
 	if u.keyboardHidFile != nil {
 		u.keyboardHidFile.Close()
```
By design.

IDisposable commented on hidrpc.go:
```diff
@@ -8,98 +8,108 @@ import (
-	case <-r:
-		scopedLogger.Debug().Dur("duration", time.Since(t)).Msg("HID RPC message handled")
+		logger.Warn().Msg("HID RPC message took too long")
+	case err := <-r:
+		logger.Debug().Dur("duration", time.Since(t)).Msg("HID RPC message handled")
+		if err != nil {
+			logger.Warn().Err(err.(error)).Msg("failed to handle HID RPC message")
```
err can only be an error and we know it's not nil here.

IDisposable commented on usb_mass_storage.go:
```diff
@@ -314,23 +333,23 @@ func rpcMountWithStorage(filename string, mode VirtualMediaMode) error {
 		return fmt.Errorf("another virtual media is already mounted")
 	}
 
 	fullPath := filepath.Join(imagesFolder, filename)
 	fileInfo, err := os.Stat(fullPath)
 	if err != nil {
-		return fmt.Errorf("failed to get file info: %w", err)
+		return err
 	}
```
Fixed

IDisposable commented on internal/usbgadget/utils.go:
```diff
@@ -112,67 +90,31 @@ func compareFileContent(oldContent []byte, newContent []byte, looserMatch bool)
 	n, err = file.Write(data)
 	if err == nil {
-		return
+		u.resetLogSuppressionCounter("writeWithTimeout_" + fileName)
```
This is not true

adamshiervani commented on jsonrpc.go:
```diff
@@ -56,7 +56,7 @@ type BacklightSettings struct {
+		jsonRpcLogger.Error().Err(err).Msg("Error marshalling JSONRPC response")
 		return
 	}
 	err = session.RPCChannel.SendText(string(responseBytes))
```
Found a nil pointer dereference that crashes the app and triggers failsafe:
```
panic: runtime error: invalid memory address or nil pointer dereferencegithub.com/pion/webrtc/v4.(*DataChannel).SendText(0x0, ...)github.com/jetkvm/kvm.writeJSONRPCResponse(...)
```

`shutdownSession()` closes rpcQueue and then sets RPCChannel = nil, but onRPCMessage is spawned in a separate goroutine (`go onRPCMessage(msg, s)`). When that goroutine finishes and calls `writeJSONRPCResponse()`, RPCChannel is already nil.

**Fix:** Add nil check to `writeJSONRPCResponse()` (same pattern as writeJSONRPCEvent already has):

IDisposable replied:
Fixed that, thanks! Also added a check for the channel being open, and used the same logic for both `writeJSONRPCReponse` and `writeJSONRPCEvent`

IDisposable commented on jsonrpc.go:
```diff
@@ -327,23 +335,21 @@ type SSHKeyState struct {
-		if !os.IsNotExist(err) {
-			return DevModeState{}, fmt.Errorf("error checking dev mode file: %w", err)
-		}
+	if _, err := os.Stat(devModeFile); err != nil && !os.IsNotExist(err) {
+		return DevModeState{}, fmt.Errorf("error checking dev mode file: %w", err)
 	} else {
-		devModeEnabled = true
+		return DevModeState{
+			Enabled: err == nil,
+		}, nil
 	}
```
There's really not much better here. If the err == nil, then things are enabled... otherwise the err _is_ a file-not-exist, which **means** that debug mode is NOT enabled.

adamshiervani reviewed (commented):
LED state synchronization stops working after 1-2 runs of `make test_e2e`. The keyboard and mouse movement both work, it's just the LED state that stops updating. A restart fixes the issue.

Also, worth mentioning is that I can also reproduce this issue latest `dev` without this PR.

IDisposable commented:
I will look into the LED state issue. When you say it "stops working" do you mean that the client doesn't track the server state, or do you mean the UI doesn't track the RPC pushes?

adamshiervani commented:
The remote host receives all keys and mouse events, it's just that the UI doesn't properly reflect the LED State. This makes the `make test_e2e` fail, as it uses the LED State to determine whether the remote host handles keyboard presses.

IDisposable commented:
> The remote host receives all keys and mouse events, it's just that the UI doesn't properly reflect the LED State. This makes the `make test_e2e` fail, as it uses the LED State to determine whether the remote host handles keyboard presses.

I'll dig in this weekend... what LED state are we tracking, and what is the test JetKVM physically attached to on the USB port? I notice that MacOS doesn't light the CapsLock LED as I would expect (never really has...)  Do you know what kind of machine the Jet is driving?

adamshiervani commented:
The remote host i'm testing on is on Ubuntu 24.04.

IDisposable committed:
Fix gadget when attached after power-up

IDisposable committed:
Increase LED/key-down state reliability

IDisposable commented:
So I did a few things:
- Switched the push-based LED state reporting to follow the same pattern as the pushed key-down state reporting to use a channel so that we won't block when trying to send back the LED state
- Added an `OverwriteChan` _go_ `chan` wrapper that discards the oldest entry when asked to queue something and the channel is full
- Switch **both** LED and key-down state reporting channels to use the `OverwriteChan` such that when we start a session, the channels get initialized with a limit of 4 (for LED) and 16 (for key-down) pending messages... if we have more backed up, the _oldest one(s)_ get discarded and the channel is never full. I suspect that both of those limits are **much** higher than needed (probably only need most recent one).
- Updated the _devices.$id.tsx_ to always ask for LED and key-down state at route load, and setup a 30-second polling for state in case we miss the state being pushed.
- Removed the LED "Shift" state as that was not reliable across machines

adamshiervani commented:
> Updated the devices.$id.tsx to always ask for LED and key-down state at route load, and setup a 30-second polling for state in case we miss the state being pushed.

This feels like something is broken. Any special reason for this?

IDisposable commented:
>  > Updated the devices.$id.tsx to always ask for LED and key-down state at route load, and setup a 30-second polling for state in case we miss the state being pushed.
> 
> This feels like something is broken. Any special reason for this?

Honestly, this was to address your observation that we were failing occasionally to track the LED state in E2E. I didn't really ever see that, but put this in as a backup plan. We could not do it at all, or change the poll rate.

IDisposable committed:
Fix merge

adamshiervani commented:
This does not resolve issue #128 for me on an Ubuntu 24 remote host.

Steps to reproduce:
1. Start the JetKVM without a USB cable connected, powering it only via the DC extensions through the RJ-12 port.
2. Wait for the device to finish booting.
3. Plug in the USB cable.

Result:
The USB status remains “Disconnected,” and both the keyboard and mouse fail to function.

IDisposable commented:
I'm powering from the USB splitter (power side going to a dumb USB charger) plugging the USB data-side of the splitter into a MacBook after it and the JetKVM are fully booted. Shows the USB connected and the mouse and keyboard work without power any settings/reboots.  So strange.

adamshiervani commented:
That's so odd. Do you have an DC or ATX extension available to test my setup?


Also, I'm getting a null pointer when disabling all the USB Devices.
```2026-01-06T13:35:00Z ERR jetkvm jsonrpc Recovered from panic in RPC handler error="runtime error: invalid memory address or nil pointer dereference" id=12 length=157 method=setUsbDevices params={"devices":{"absolute_mouse":false,"keyboard":false,"mass_storage":false,"relative_mouse":false}}
2026-01-06T13:35:00Z ERR jetkvm jsonrpc Error calling RPC handler error="runtime error: invalid memory address or nil pointer dereference" id=12 length=157 method=setUsbDevices params={"devices":{"absolute_mouse":false,"keyboard":false,"mass_storage":false,"relative_mouse":false}}
2026-01-06T13:35:02Z INF jetkvm gin Request body_size=16 ip=192.168.1.40 latency=0.1785 method=GET path=/device/status status=200 user_agent="Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/143.0.0.0 Safari/537.36"```

Once fixed, could you also make an e2e test for that "disable all USB devices" case?

MadYeti1984 commented:
I just got around to testing my JetKVM. I've tested on 2 Windows 11 machines (latest updates). I cannot get the USB to register as connected on any USB ports on either machine. I made sure to update the JetKVM software first. Tried power cycling both PCs and JetKVM in different order, and still nothing. So, as of now can't use the KM rendering device unusable.

IDisposable commented:
Use the front panel to reset the JetKVM

On Fri, Jan 16, 2026 at 1:26 PM MadYeti1984 ***@***.***>
wrote:

> *MadYeti1984* left a comment (jetkvm/kvm#1062)
> <https://github.com/jetkvm/kvm/pull/1062#issuecomment-3761475285>
>
> I just got around to testing my JetKVM. I've tested on 2 Windows 11
> machines (latest updates). I cannot get the USB to register as connected on
> any USB ports on either machine. I made sure to update the JetKVM software
> first. Tried power cycling both PCs and JetKVM in different order, and
> still nothing. So, as of now can't use the KM rendering device unusable.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/jetkvm/kvm/pull/1062#issuecomment-3761475285>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAGVMS42JJARQLEZZNFAKMT4HE3NXAVCNFSM6AAAAACOX6O6WKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTONRRGQ3TKMRYGU>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Marc Brooks
http://about.me/IDisposable
http://twitter.com/IDisposable

MadYeti1984 commented:
Hey, I was able to get this to work finally- the solution was to use a
different USB-C to USB-A cable. Neither of the cables provided (USBC-USBC
or USBC-USBA) worked.

On Fri, Jan 16, 2026 at 5:53 PM Marc Brooks ***@***.***>
wrote:

> *IDisposable* left a comment (jetkvm/kvm#1062)
> <https://github.com/jetkvm/kvm/pull/1062#issuecomment-3762246039>
> Use the front panel to reset the JetKVM
>
> On Fri, Jan 16, 2026 at 1:26 PM MadYeti1984 ***@***.***>
> wrote:
>
> > *MadYeti1984* left a comment (jetkvm/kvm#1062)
> > <https://github.com/jetkvm/kvm/pull/1062#issuecomment-3761475285>
> >
> > I just got around to testing my JetKVM. I've tested on 2 Windows 11
> > machines (latest updates). I cannot get the USB to register as connected
> on
> > any USB ports on either machine. I made sure to update the JetKVM
> software
> > first. Tried power cycling both PCs and JetKVM in different order, and
> > still nothing. So, as of now can't use the KM rendering device unusable.
> >
> > —
> > Reply to this email directly, view it on GitHub
> > <https://github.com/jetkvm/kvm/pull/1062#issuecomment-3761475285>, or
> > unsubscribe
> > <
> https://github.com/notifications/unsubscribe-auth/AAGVMS42JJARQLEZZNFAKMT4HE3NXAVCNFSM6AAAAACOX6O6WKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTONRRGQ3TKMRYGU>
>
> > .
> > You are receiving this because you authored the thread.Message ID:
> > ***@***.***>
> >
>
>
> --
> Marc Brooks
> http://about.me/IDisposable
> http://twitter.com/IDisposable
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/jetkvm/kvm/pull/1062#issuecomment-3762246039>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A6KFICHSBYCR5RYS5KJG3234HF2YTAVCNFSM6AAAAACOX6O6WKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTONRSGI2DMMBTHE>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>
